src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs

```diff
@@ -30,6 +30,244 @@ public static IMethodSymbol FindLocalFunction(this CommonTestBase.CompilationVer
+        Console.WriteLine(a());
+    }
+
+    public static void Main(string[] args)
+    {
+        var c = new C();
+        c.M();
+    }
+}";
+            VerifyOutput(src, "111");
+
+        }
+
+        [Fact]
+        public void LocalFuncAndLambdaWithDifferentThis4()
+        {
+            var src = @"
+using System;
+class C
+{
+    private int P = 1;
+    public void M()
+    {
+        int Local(int x) => x + this.P;
+
+        int y = 10;
+        var a = new Func<int>(() =>
+        {
+            var b = (Func<int, int>)Local;
+            return b(y);
+        });
+        Console.WriteLine(a());
+    }
+
+    public static void Main(string[] args)
+    {
+        var c = new C();
+        c.M();
+    }
+}";
+            VerifyOutput(src, "11");
+        }
+
+        [Fact]
+        public void LocalFuncAndLambdaWithDifferentThis5()
+        {
+            var src = @"
+using System;
+class C
+{
+    private int P = 1;
+    public void M()
+    {
+        int Local(int x) => x + this.P;
+
+        int y = 10;
+        var a = new Func<int>(() =>
+        {
+            var b = new Func<int, int>(Local);
+            return b(y);
+        });
+        Console.WriteLine(a());
+    }
+
+    public static void Main(string[] args)
+    {
+        var c = new C();
+        c.M();
+    }
+}";
+            VerifyOutput(src, "11");
+        }
+
+        [Fact]
+        public void TwoFrames()
+        {
+            var src = @"
+using System;
+class C
+{
+    private int P = 0;
+    public void M()
+    {
+        int x = 0;
+
+        var a = new Func<int>(() =>
+        {
+            int Local() => x + this.P;
+            int z = 0;
+            int Local3() => z + Local();
+            return Local3();
+        });
+        Console.WriteLine(a());
+    }
+
+    public static void Main(string[] args)
+    {
+        var c = new C();
+        c.M();
+    }
+}";
+            VerifyOutput(src, "0");
+        }
+
+        [Fact]
+        public void SameFrame()
+        {
+            var src = @"
+using System;
+class C
+{
+    private int P = 1;
+    public void M()
+    {
+        int x = 10;
+        int Local() => x + this.P;
+
+        int y = 100;
+        int Local2() => y + Local();
+        Console.WriteLine(Local2());
+    }
+
+    public static void Main(string[] args)
+    {
+        var c = new C();
+        c.M();
+    }
+}";
+            VerifyOutput(src, "111");
+        }
+
+        [Fact]
+        public void MutuallyRecursiveThisCapture()
+        {
+            var src = @"
+using System;
+class C
+{
+    private int P = 1;
+    public void M()
+    {
+        int Local()
+        {
+            if (this.P < 5)
+            {
+                return Local2(this.P++);
+            }
+            else
+            {
+                return 1;
+            }
+        }
+        int Local2(int x) => x + Local();
+        Console.WriteLine(Local());
+    }
+    public static void Main() => new C().M();
+}";
```

Comment:
I'd add more scenarios for accessing same local function from different levels of frame nesting.  
With variations when the function does not lift, lift only "this" or lift some local

Something like the following (with or without "capturedX" locals:

'''cs
{
    C1 LocalFunc1() => this;   // can also lift a local or nothing at all

    {
        C1 LocalFunc2() => LocalFunc1();
        
        int captured1 = 1;
        Func<int> f = ()=> captured1 + LocalFunc2().GetHashCode();;
        f();

        {
            C1 LocalFunc3() => LoclaFunc2();

            int captured2 = 1;
            Func<int> f1 = ()=> captured1 + captured2 + LocalFunc3().GetHashCode();;
            f2();

            {
                    // only needs "this"
                    C1 LocalFunc4() => LocalFunc3();

                    int captured3 = 1;

                    // needs a chain of proxies to fetch "this"
                    Func<int> f3 = ()=> captured1 + captured2 + captured3 + LocalFunc4().GetHashCode();
                    f3();

                     // can get to "this" directly.
                     WriteLine(LocalFunc4());
             }
        }
     }
}
```